### DISCUSSION TITLE
dbus-serialbattery with hybrid inverter

### DISCUSSION THREAD

**Discussion by nobuoku77** [Louisvdw/dbus-serialbattery#338] [Q&A] [answered]
I now have 4 USB adapters(1 hybrid inverter, 2 jk-BMS and 1 Wi-Sun adapter) connected to my raspberry pi to monitor data. Please let me know how to pass this data to venus OS. I thought that I should use this dbus-serialbattery for jk-BMS, so I tried serial communication with the investor under that environment, but it didn't work. Are you aware that other serial communication is not possible while this dbus-serialbattery is running?
If I have to write a driver for venus OS from scratch, is there a program I should refer to?
![hems2](https://user-images.githubusercontent.com/120176759/206853482-b88fc3f9-1392-4259-93eb-5832c3a32195.jpg)

**Comment by Louisvdw** [comment, marked as answer]
You might want to try something like https://github.com/syssi/esphome-jk-bms
dbus-serialbattery only work with Victron Energy's VenusOS, and while you can use that as a stand alone set up as well other inverters do not interface with VenusOS that I have seen. 

Serial communication is always between a master and a slave. RS485 does support mulitple slaves and one master, but then you need to address each slave with it's own slave. Most BMS does not have different address availabe in their protocol, including JKBMS, so you can only talk between 1 master and 1 slave.

**Reply by nobuoku77** [reply]
Yes, I read esphome-jk-bms and other programs and made a standalone system. My next goal is to make the venus OS display the inverter(SRNE  ASF series) data, and I'm looking for a clue to write a driver for that.
About serial communication, I prepare each USB port for each device and connect one to one (/dev/ttyUSB0 - /dev/ttyUSB4)
What confused me was that when I tried to run a python program using /dev/ttyUSB2 on VenusOS running dbus-serialbattery which use /dev/ttyUSB0 and /dev/ttyUSB1 for 2 JK-BMS, I couldn't communicate.
If dbus-serialbattery manages USB ports all together because it checks for new insertions into USB ports, is it possible to extend dbus-serialbattery into a program like dbus-serialdevice?

**Reply by Louisvdw** [reply]
The VenusOS system has a program called serial-starter which reads any new connected port and tries to start drivers for that. This will in most cases cause a clash if you also want to connect to that port, so you need to first stop serial-starter for that port.

So if I want to manually run the driver on port ttyUSB0, I first need to stop serialstarter for that port like so:
> /opt/victronenergy/serial-starter/stop-tty.sh /dev/ttyUSB0

There is also a start-tty.sh if you want to start it later.
Check the Victron github account on github.com for all the documentation.

**Comment by nobuoku77** [comment]
Thank you . I didn't know where to start from the huge amount of documentation, but thanks to your advice, I think I can get started.
